strax/processing/hitlets.py

```diff
@@ -133,219 +136,98 @@ def refresh_hit_to_hitlets(hits, hitlets):
     """
-    for ind, hit in enumerate(new_hitlets):
-
-        updated = False
-        counter = 0
-        current_ri = hit['record_i']
-        while not updated:
-            r = records[current_ri]
-            # Hitlet must only partially be contained in record_i:
-            time = hit['time']
-            end_time = strax.endtime(hit)
-            start_in = (r['time'] <= time) & (time < strax.endtime(r))
-            end_in = (r['time'] < end_time) & (end_time <= strax.endtime(r))
-            if start_in or end_in:
-                hit['record_i'] = current_ri
-                break
-            else:
-                last_ri = current_ri
-                current_ri = next_ri[current_ri]
-                counter += 1
-                
-            if current_ri == -1:
-                print('Record:\n', r, '\nHit:\n', hit)
-                raise ValueError('Was not able to find record_i')
-
-            if counter > TRIAL_COUNTER_NEIGHBORING_RECORDS:
-                print(ind, last_ri)
-                raise RuntimeError('Tried too often to find correct record_i.')
+    # Numba will not raise any exceptions if to_pe is too short, leading
+    # to strange bugs. This check is somewhat not complete, but should
+    # be sufficient in most cases.
+    to_pe_has_wrong_shape = len(to_pe) < hitlets['channel'].max()
+    if to_pe_has_wrong_shape:
+        raise ValueError('"to_pe" has a wrong shape. Array index must'
+                         ' match channel numbers.')
+
+    hitelts_is_single_row = isinstance(hitlets, np.void)
+    if hitelts_is_single_row:
+        # A structured array becomes void type if a single row is called,
+        # e.g. hitlets[0] which does not work in numba while, hitlets[:1]
+        # does. So we have to convert the row into the correct format first.
+        hitlets = np.array([hitlets])
+
+    data_field_in_hitlets = 'data' in hitlets.dtype.names
+    if data_field_in_hitlets:
+        data_is_not_empty = np.any(hitlets['data'] != 0)
+        if data_is_not_empty:
+            raise ValueError('The data field of hitlets must be empty!')
+
+        data_field_not_long_enough = len(hitlets[0]['data']) < hitlets['length'].max()
+        if data_field_not_long_enough:
+            raise ValueError('The data field must be as large as the longest hitlet in our data.')
+
+        hitlets_with_data_field = hitlets
+    else:
+        n_samples = max(100, hitlets['length'].max())
+        hitlets_with_data_field = np.zeros(len(hitlets), strax.hitlet_with_data_dtype(n_samples))
+        strax.copy_to_buffer(hitlets,
+                             hitlets_with_data_field,
+                             '_copy_hitlets_to_hitlets_width_data')
+
+    _get_hitlets_data(hitlets_with_data_field, records, to_pe)
+    return hitlets_with_data_field
+
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _get_hitlets_data(hitlets, records, to_pe):
+    rranges = _touching_windows(records['time'],
```

Comment:
You mean `if r['channel'] != h['channel']:`, so this solves all problems because hitlets are fully inclusive of records and therefore a touching window can only yield touching one record as the boundaries of a hitlet are always broader/equal to records start/end time with some extra space left and right?

Reply:
I am confused by your question "hitlets are fully inclusive of records" and "hitlet are always broader/equal to records start/end". These two statements are the opposite of another is not it? But either way let me try to clarify: 
A hitlet can in general touch multiple record fragments. A hitlet does need to be fully contained in a record due to the left/right hit extension. As fragments are sorted by time and channel this means that in principle a hitlet can also touch the fragments of another channel. Hence Tianyu required  `if r['channel'] != h['channel']:` to solve this problem.